### ISSUE TITLE
const_array is absurdly slow

### ISSUE THREAD

**Issue by Zaharid** [numba/numba#875]
Consider the following:

``` python
import numba
import numpy as np
x = np.random.rand(100,100,100,100)
@numba.jit(nopython = True, locals={'x':'f8[:,:,:,:]'})
def nose(y):
   y += x[0,0,0,0]
nose(1)
```

Even declaring explicitly the type,  numba seems to python-cycle though every element in the array (the `get_constant` function).
The equivalent:

``` python
@numba.jit(nopython = True)
def nose2(x,y):
   y += x[0,0,0,0]
```

works fine.

**Comment by pitrou** [comment]
Confirmed on 0.23.0.

**Comment by ma-ba** [comment]
PR #1747 also resolved a bug in Numba <= 0.24.0 for which `make_constant_array` would choke on empty arrays.
```python
from numba import njit
import numpy as np

a = np.empty(0, dtype=np.int32)
@njit
def get_element(i):
    if a.shape[0] > i:
        return a[i]
    else:
        return -1

print(get_element(0))
```
```
Traceback (most recent call last):
  File "lib/python3.4/site-packages/numba/lowering.py", line 196, in lower_block
    self.lower_inst(inst)
  File "lib/python3.4/site-packages/numba/lowering.py", line 238, in lower_inst
    val = self.lower_assign(ty, inst)
  File "lib/python3.4/site-packages/numba/lowering.py", line 375, in lower_assign
    value.value)
  File "lib/python3.4/site-packages/numba/targets/base.py", line 364, in get_constant_generic
    return self.make_constant_array(builder, ty, val)
  File "lib/python3.4/site-packages/numba/targets/base.py", line 832, in make_constant_array
    lldtype = values[0].type
IndexError: list index out of range

During handling of the above exception, another exception occurred:
...
```
Just an FYI. Works perfectly fine with the current development version!
